website_product_subscription/models/subscription_request.py

```diff
@@ -25,32 +34,57 @@ def send_gift_emails(self):
             ".gift_subscription_existing_user_mail_template"
         )
         for request in self:
-            if self.env["res.users"].user_exists(request.subscriber.email):
-                existing_user_template.send_mail(request.id)
+            if (
+                request.gift_sent
+                or request.type != "gift"
+                or request.gift_date > fields.Date.today()
+            ):
+                # Don't send an e-mail (again, at all, or yet, in that order).
+                continue
+            if not request.subscriber.email:
+                _logger.error(
+                    "partner %s %s does not have an email address; cannot send gift email"
+                    % (request.subscriber.id, request.subscriber.name)
+                )
+                continue
+            if request.subscriber._has_web_access():
+                template = existing_user_template
             else:
-                new_user_template.send_mail(request.id)
-                request.subscriber.create_web_access()
+                template = new_user_template
+            template.send_mail(request.id)
             request.gift_sent = True
 
     @api.multi
-    def create_web_access(self):
-        today = fields.Date.today()
+    def create_subscriber_web_access(self):
         for request in self:
-            if request.type == "gift" and request.gift_date <= today:
-                request.send_gift_emails()
-            elif request.type == "gift" and request.gift_date > today:
-                continue
-            else:
-                request.subscriber.create_web_access()
+            request.subscriber.create_web_access()
 
     @api.model
     def cron_create_scheduled_gift_user(self):
-        today = fields.Date.today()
+        now = fields.Datetime.from_string(fields.Datetime.now()).time()
+        if now < datetime.time(7, 0, 0) or now > datetime.time(22, 0, 0):
+            # Don't send e-mails during hours meant for sleep.
+            _logger.info(
+                "Skipping cron_create_scheduled_gift_user because it's late: %s"
+                % now
+            )
+            return
+
         requests = self.search(
             [
                 ("gift", "=", True),
                 ("gift_sent", "=", False),
-                ("gift_date", "<=", today),
+                # TODO: Revert following back to '<=' once the back log has
+                # cleared. (T3833)
+                ("gift_date", "=", fields.Date.today()),
             ]
         )
```

Comment:
Should the following be added?

```python
("state", "=", "paid")
```

or something like

```python
("state", "!=", "cancel")
```

The states are `["draft", "sent", "paid", "cancel"]`

Reply:
This is a question for Wilfried or Medor, I'd say `("state", "in", ("sent", "paid"))`.

Reply:
I can't really imagine any scenario where cancelled requests should receive e-emails. Asked Cath about this.

Reply:
sure but I'm not so sure about "draft" requests.

Reply:
Talked with Cath; we'll keep this as-is and cheat by setting `gift_sent` to True for people in the back log who will not receive e-mails.